Race condition

Ref: https://github.com/plane-watch/pw-slippymap/pull/9/commits/3798109d5af0437021204e5fe24e770b0a9e67b6

When running `go test` with `-race`, the following race condition appears to occur:

```
Command: go test -race -v ./...
[11](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:11)
/usr/bin/xvfb-run --auto-servernum go test -race -v ./...
[12](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:12)
go: downloading github.com/hajimehoshi/ebiten/v2 v2.2.7
[13](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:13)
go: downloading github.com/jezek/xgb v0.0.0-20210312150743-0e0f116e1240
[14](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:14)
go: downloading github.com/go-gl/glfw/v3.3/glfw v0.0.0-20210727001814-0db043d8d5be
[15](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:15)
?   	pw_slippymap	[no test files]
[16](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:16)
?   	pw_slippymap/localdata	[no test files]
[17](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:17)
?   	pw_slippymap/markers	[no test files]
[18](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:18)
=== RUN   TestNewCachedTileProvider
[19](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:19)
--- PASS: TestNewCachedTileProvider (0.24s)
[20](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:20)
=== RUN   TestGetTileAddress
[21](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:21)
--- PASS: TestGetTileAddress (0.00s)
[22](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:22)
=== RUN   TestSlippyMap
[23](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:23)
User home dir: /home/runner
[24](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:24)
2022/04/18 07:20:13 /home/runner/.plane.watch does not exist, creating
[25](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:25)
2022/04/18 07:20:13 /home/runner/.plane.watch/tilecache does not exist, creating
[26](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:26)
==================
[27](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:27)
WARNING: DATA RACE
[28](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:28)
Read at 0x00c000018628 by goroutine 27:
[29](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:29)
  pw_slippymap/slippymap.(*OSMTileProvider).GetTileAddress()
[30](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:30)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/osmtileprovider.go:20 +0x58
[31](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:31)
  pw_slippymap/slippymap.(*CachedTileProvider).GetTileAddress()
[32](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:32)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/cachedtileprovider.go:54 +0x23d
[33](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:33)
  pw_slippymap/slippymap.(*SlippyMap).makeTile.func1()
[34](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:34)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:290 +0xbe
[35](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:35)

[36](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:36)
Previous write at 0x00c000018628 by goroutine 26:
[37](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:37)
  pw_slippymap/slippymap.(*OSMTileProvider).GetTileAddress()
[38](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:38)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/osmtileprovider.go:26 +0x25b
[39](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:39)
  pw_slippymap/slippymap.(*CachedTileProvider).GetTileAddress()
[40](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:40)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/cachedtileprovider.go:54 +0x23d
[41](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:41)
  pw_slippymap/slippymap.(*SlippyMap).makeTile.func1()
[42](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:42)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:290 +0xbe
[43](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:43)

[44](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:44)
Goroutine 27 (running) created at:
[45](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:45)
  pw_slippymap/slippymap.(*SlippyMap).makeTile()
[46](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:46)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:288 +0x27c
[47](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:47)
  pw_slippymap/slippymap.(*SlippyMap).makeTileToTheLeft()
[48](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:48)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:219 +0x2cc
[49](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:49)
  pw_slippymap/slippymap.(*SlippyMap).Update()
[50](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:50)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:125 +0x524
[51](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:51)
  pw_slippymap/slippymap.NewSlippyMap()
[52](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:52)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:460 +0x2aa
[53](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:53)
  pw_slippymap/slippymap.TestSlippyMap()
[54](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:54)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap_test.go:23 +0x15a
[55](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:55)
  testing.tRunner()
[56](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:56)
      /opt/hostedtoolcache/go/1.17.8/x64/src/testing/testing.go:1259 +0x22f
[57](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:57)
  testing.(*T).Run·dwrap·21()
[58](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:58)
      /opt/hostedtoolcache/go/1.17.8/x64/src/testing/testing.go:1306 +0x47
[59](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:59)

[60](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:60)
Goroutine 26 (running) created at:
[61](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:61)
  pw_slippymap/slippymap.(*SlippyMap).makeTile()
[62](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:62)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:288 +0x27c
[63](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:63)
  pw_slippymap/slippymap.(*SlippyMap).makeTileAbove()
[64](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:64)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:161 +0x2b8
[65](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:65)
  pw_slippymap/slippymap.(*SlippyMap).Update()
[66](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:66)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:122 +0x509
[67](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:67)
  pw_slippymap/slippymap.NewSlippyMap()
[68](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:68)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap.go:460 +0x2aa
[69](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:69)
  pw_slippymap/slippymap.TestSlippyMap()
[70](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:70)
      /home/runner/work/pw-slippymap/pw-slippymap/slippymap/slippymap_test.go:23 +0x15a
[71](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:71)
  testing.tRunner()
[72](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:72)
      /opt/hostedtoolcache/go/1.17.8/x64/src/testing/testing.go:1259 +0x22f
[73](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:73)
  testing.(*T).Run·dwrap·21()
[74](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:74)
      /opt/hostedtoolcache/go/1.17.8/x64/src/testing/testing.go:1306 +0x47
[75](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:75)
==================
[76](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:76)
    testing.go:1152: race detected during execution of test
[77](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:77)
--- FAIL: TestSlippyMap (0.16s)
[78](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:78)
=== RUN   TestGpsCoordsToTileInfo
[79](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:79)
--- PASS: TestGpsCoordsToTileInfo (0.00s)
[80](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:80)
=== RUN   TestTileXYZtoGpsCoords
[81](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:81)
--- PASS: TestTileXYZtoGpsCoords (0.00s)
[82](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:82)
=== CONT  
[83](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:83)
    testing.go:1152: race detected during execution of test
[84](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:84)
FAIL
[85](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:85)
FAIL	pw_slippymap/slippymap	1.705s
[86](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:86)
FAIL
[87](https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true#step:5:87)
```

This only seems to happen on Github's runners - I can't replicate the problem locally...

Relevant test: https://github.com/plane-watch/pw-slippymap/runs/6059849270?check_suite_focus=true

Is this an issue we need to look into?

--- a/go.mod
+++ b/go.mod
@@ -5,11 +5,15 @@ go 1.17
 require github.com/hajimehoshi/ebiten/v2 v2.2.7
 
 require (
+	github.com/davecgh/go-spew v1.1.0 // indirect
 	github.com/go-gl/glfw/v3.3/glfw v0.0.0-20210727001814-0db043d8d5be // indirect
 	github.com/jezek/xgb v0.0.0-20210312150743-0e0f116e1240 // indirect
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/stretchr/testify v1.7.1 // indirect
 	golang.org/x/exp v0.0.0-20190731235908-ec7cb31e5a56 // indirect
 	golang.org/x/image v0.0.0-20210628002857-a66eb6448b8d // indirect
 	golang.org/x/mobile v0.0.0-20210902104108-5d9a33257ab5 // indirect
 	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c // indirect
 	golang.org/x/sys v0.0.0-20211019181941-9d821ace8654 // indirect
+	gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c // indirect
 )

--- a/slippymap/cachedtileprovider_test.go
+++ b/slippymap/cachedtileprovider_test.go
@@ -1,6 +1,7 @@
 package slippymap
 
 import (
+	"errors"
 	"io/ioutil"
 	"os"
 	"path"
@@ -39,13 +40,8 @@ func TestNewCachedTileProvider(t *testing.T) {
 		t.Errorf("Expected: %s, got: %s", expectedPath, tilePath)
 	}
 
-	// prepare "faulty" OSMTileProvider
-	faultyOSMTileProvider := OSMTileProvider{
-		osm_url_prefix: 4,
-	}
-
 	// get cached tile provider
-	ctp = NewCachedTileProvider(dir, &faultyOSMTileProvider)
+	ctp = NewCachedTileProvider(dir, &FaultyTileProvider{})
 
 	// request a tile
 	tilePath, err = ctp.GetTileAddress(2, 3, 4)
@@ -55,4 +51,10 @@ func TestNewCachedTileProvider(t *testing.T) {
 		t.Errorf("Expected an error, got none")
 	}
 
+}
+
+type FaultyTileProvider struct{}
+
+func (FaultyTileProvider) GetTileAddress(int, int, int) (string, error) {
+	return "", errors.New("oh no an error")
 }

--- a/slippymap/osmtileprovider.go
+++ b/slippymap/osmtileprovider.go
@@ -3,12 +3,13 @@ package slippymap
 import (
 	"errors"
 	"fmt"
+	"sync/atomic"
 )
 
 // OSMTileProvider generates the URLs to OSM tiles
 type OSMTileProvider struct {
 	// used to round-robin connections to OSM servers, as-per https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Tile_servers
-	osm_url_prefix int
+	osm_url_prefix int32
 }
 
 var _ TileProvider = &OSMTileProvider{}
@@ -17,16 +18,22 @@ func (op *OSMTileProvider) GetTileAddress(x, y, z int) (tilePath string, err err
 	var url string
 	// returns URL to open street map tile
 	// load balance urls across servers as-per OSM guidelines: https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Tile_servers
-	switch op.osm_url_prefix {
+
+	// we don't really care about atomically incrementing this number, but golang race detection bitches
+	// at us if we let this be subject to race conditions. Performance not super critical so just use
+	// atomic to make the race detector shut up
+	nextCDN := atomic.AddInt32(&op.osm_url_prefix, 1) % 3
+
+	switch nextCDN {
 	case 0:
 		url = fmt.Sprintf("http://a.tile.openstreetmap.org/%d/%d/%d.png", z, x, y)
-		op.osm_url_prefix = 1
+
 	case 1:
 		url = fmt.Sprintf("http://b.tile.openstreetmap.org/%d/%d/%d.png", z, x, y)
-		op.osm_url_prefix = 2
+
 	case 2:
 		url = fmt.Sprintf("http://c.tile.openstreetmap.org/%d/%d/%d.png", z, x, y)
-		op.osm_url_prefix = 0
+
 	default:
 		return "", errors.New("invalid osm_url_prefix")
 	}

--- a/slippymap/osmtileprovider_test.go
+++ b/slippymap/osmtileprovider_test.go
@@ -1,53 +1,36 @@
 package slippymap
 
-import "testing"
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
 
 func TestGetTileAddress(t *testing.T) {
 
-	TileProvider := &OSMTileProvider{}
+	t.Run("GetTileAddress loops through CDNs correctly", func(t *testing.T) {
+		// -1 so that the first increment lands us up on CDN A. Doesn't really matter where it starts though
+		TileProvider := &OSMTileProvider{osm_url_prefix: -1}
 
-	// check correct a... url is returned
-	url, err := TileProvider.GetTileAddress(1, 2, 3)
-	if err != nil {
-		t.Error(err)
-	}
-	if url != "http://a.tile.openstreetmap.org/3/1/2.png" {
-		t.Errorf("Expected http://a.tile.openstreetmap.org/3/1/2.png, got: %s", url)
-	}
+		// check correct a... url is returned
+		url, err := TileProvider.GetTileAddress(1, 2, 3)
+		require.NoError(t, err)
+		assert.Equal(t, "http://a.tile.openstreetmap.org/3/1/2.png", url)
 
-	// check correct b... url is returned
-	url, err = TileProvider.GetTileAddress(1, 2, 3)
-	if err != nil {
-		t.Error(err)
-	}
-	if url != "http://b.tile.openstreetmap.org/3/1/2.png" {
-		t.Errorf("Expected http://b.tile.openstreetmap.org/3/1/2.png, got: %s", url)
-	}
+		// check correct b... url is returned
+		url, err = TileProvider.GetTileAddress(1, 2, 3)
+		require.NoError(t, err)
+		assert.Equal(t, "http://b.tile.openstreetmap.org/3/1/2.png", url)
 
-	// check correct c... url is returned
-	url, err = TileProvider.GetTileAddress(1, 2, 3)
-	if err != nil {
-		t.Error(err)
-	}
-	if url != "http://c.tile.openstreetmap.org/3/1/2.png" {
-		t.Errorf("Expected http://c.tile.openstreetmap.org/3/1/2.png, got: %s", url)
-	}
+		// check correct c... url is returned
+		url, err = TileProvider.GetTileAddress(1, 2, 3)
+		require.NoError(t, err)
+		assert.Equal(t, "http://c.tile.openstreetmap.org/3/1/2.png", url)
 
-	// check correct a... url is returned (it loops back to a after c)
-	url, err = TileProvider.GetTileAddress(1, 2, 3)
-	if err != nil {
-		t.Error(err)
-	}
-	if url != "http://a.tile.openstreetmap.org/3/1/2.png" {
-		t.Errorf("Expected http://a.tile.openstreetmap.org/3/1/2.png, got: %s", url)
-	}
-
-	// set invalid tile prefix to generate an error
-	TileProvider.osm_url_prefix = 4
-	url, err = TileProvider.GetTileAddress(1, 2, 3)
-	if err != nil {
-		// test passes
-	} else {
-		t.Error("Expected an error, got none.")
-	}
+		// check correct a... url is returned (it loops back to a after c)
+		url, err = TileProvider.GetTileAddress(1, 2, 3)
+		require.NoError(t, err)
+		assert.Equal(t, "http://a.tile.openstreetmap.org/3/1/2.png", url)
+	})
 }

